src/rt-app.c

```diff
@@ -989,93 +978,135 @@ static void set_thread_priority(thread_data_t *data, sched_data_t *sched_data)
-				   data->ind, sched_data->util_min);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_min=%d",
-				   sched_data->util_min);
-		}
-		if (sched_data->util_max != -1) {
-			sa_params.sched_util_max = sched_data->util_max;
-			sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MAX;
-			log_notice("[%d] setting util_max=%d",
-				   data->ind, sched_data->util_max);
-			log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
-				   "rtapp_attrs: event=uclamp util_max=%d",
-				   sched_data->util_max);
-		}
+	sa_params.size = sizeof(struct sched_attr);
+	sa_params.sched_flags = SCHED_FLAG_KEEP_ALL;
+	tid = gettid();
 
-		ret = sched_setattr(tid, &sa_params, flags);
-		if (ret != 0) {
-			log_critical("[%d] sched_setattr returned %d",
-				     data->ind, ret);
-			perror("sched_setattr: failed to set uclamp value(s)");
+	/*
+	 * Currently we cannot use the syscall to set a clamp
+	 * value without specifing a valid policy and priotiry.
+	 * Read the current values and use them for the following
+	 * syscall.
+	 * TODO: fix __sched_setscheduler() to get rid of this!
+	 */
+	ret = sched_getattr(tid, &sa_params,
+			    sizeof(struct sched_attr), flags);
+	if (ret != 0) {
+		log_critical("[%d] sched_getattr returned %d",
+				data->ind, ret);
+		errno = ret;
+		perror("sched_getattr");
+		exit(EXIT_FAILURE);
+	}
+
+	if (sched_data->util_min != -1) {
+		sa_params.sched_util_min = sched_data->util_min;
+		sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MIN;
+		log_notice("[%d] setting util_min=%d",
+			   data->ind, sched_data->util_min);
+		log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
+			   "rtapp_attrs: event=uclamp util_min=%d",
+			   sched_data->util_min);
+	}
+	if (sched_data->util_max != -1) {
+		sa_params.sched_util_max = sched_data->util_max;
+		sa_params.sched_flags |= SCHED_FLAG_UTIL_CLAMP_MAX;
+		log_notice("[%d] setting util_max=%d",
+			   data->ind, sched_data->util_max);
+		log_ftrace(ft_data.marker_fd, FTRACE_ATTRS,
+			   "rtapp_attrs: event=uclamp util_max=%d",
+			   sched_data->util_max);
+	}
+
+	ret = sched_setattr(tid, &sa_params, flags);
+	if (ret != 0) {
+		log_critical("[%d] sched_setattr returned %d",
+			     data->ind, ret);
+		perror("sched_setattr: failed to set uclamp value(s)");
+		exit(EXIT_FAILURE);
+	}
+}
+
+static void set_thread_param(thread_data_t *data, sched_data_t *sched_data)
+{
+	if (sched_data == NULL)
+		return;
+
+	if (data->curr_sched_data == sched_data)
+		return;
+
+	/* if policy not specificed, reuse previous policy */
+	if ((sched_data->policy == same) && (data->curr_sched_data))
+		sched_data->policy = data->curr_sched_data->policy;
+
+	switch (sched_data->policy) {
+		case rr:
+		case fifo:
+		case other:
+		case idle:
+		case deadline:
+			break;
+		default:
+			log_error("Unknown scheduling policy %d",
+				  sched_data->policy);
 			exit(EXIT_FAILURE);
-		}
 	}
 
+	_set_thread_priority(data, sched_data);
+	_set_thread_deadline(data, sched_data);
```

Comment:
deadline should be part of _set_thread_priority. There is no reason to make it special

Reply:
It was to avoid having something weird in `_set_thread_priority()`: 

```
/*
 * If THREAD_PRIORITY_UNCHANGED, but we are deadline, it's probably because we
 * want to edit one of the deadline parameters.
 */
if (sched_data->prio == THREAD_PRIORITY_UNCHANGED &&
     sched_data->policy != deadline)
          return;
```

But I can make the change if you'd rather prefer to have all the policies covered in _set_thread_priority.

Reply:
The rule is  that : if there is a sched_data, it's means that there is something to change on the scheduler policy/prio. Now,